### PR TITLE
Document OS-specific approaches for noarch packages

### PR DESCRIPTION
**Pull request by jaimergp**
PR Checklist:

- [ ] note any issues closed by this PR with [closing keywords](https://help.github.com/articles/closing-issues-using-keywords)
- [ ] put any other relevant information below

This creative approach is used by several feedstocks (10 or so, according to [this Github CodeSearch query](https://cs.github.com/?scopeName=All+repos&scope=&q=org%3Aconda-forge+path%3Aconda-forge.yml+noarch_platforms)) now but wasn't documented. This approach alleviates the number of builds needed for platform-specific dependencies in noarch-eligible packages.

### CODE REVIEW

File: src/maintainer/knowledge_base.rst
Comment (mattip) [inline src/maintainer/knowledge_base.rst:1047]:
```suggestion
1.  Virtual packages. Prefixed with a double underscore, they are used by conda
    when running the dependency resolver to represent properties of the currently
    activated environment 
```

File: src/maintainer/knowledge_base.rst
Comment (mattip) [inline src/maintainer/knowledge_base.rst:1049]:
```suggestion
    the activated environment is Linux, Windows, or MacOS, respectively. ``__unix`` is present in both Linux and MacOS. Note
```

File: src/maintainer/knowledge_base.rst
Comment (mattip) [inline src/maintainer/knowledge_base.rst]:
```suggestion

```

File: src/maintainer/knowledge_base.rst
Comment (jaimergp) [reply src/maintainer/knowledge_base.rst:1049]: I think one can only change the virtual packages with `CONDA_OVERRIDE_*` env variables. Otherwise, regardless the active environment, it will still report the native platform type.

Comment (mattip) [reply src/maintainer/knowledge_base.rst:1049]: OK, thanks for the clarification. I don't know if "running platform" is defined somewhere?

Comment (mattip) [reply src/maintainer/knowledge_base.rst:1049]: also, it is not clear to me from "as constraints for the solver" that this happens at runtime, and not while building the package.

File: src/maintainer/knowledge_base.rst
Comment (jaimergp) [reply src/maintainer/knowledge_base.rst:1047]: Same as below, technically they are environment independent unless overridden.

File: src/maintainer/knowledge_base.rst
Comment (jaimergp) [reply src/maintainer/knowledge_base.rst:1049]: I'll rephrase, thanks!

**Comment by jakirkham** [comment]
Thanks for adding this, Jaime! 🙏

cc @minrk (in case you have thoughts on this 🙂)

File: src/maintainer/knowledge_base.rst
Comment (mattip) [inline src/maintainer/knowledge_base.rst:1116]:
```suggestion
would never be true). Fortunately, we can change the default behaviour in ``./conda-forge.yml``:
```

**Review by goanpeca — APPROVED** [review]

File: src/maintainer/knowledge_base.rst
Comment (jaimergp) [reply src/maintainer/knowledge_base.rst:1102]: Ok my main concern about not using `conda_build_config.yaml`, @isuruf: 

* The package hashes will be the same (see https://github.com/conda-forge/click-feedstock/pull/44), so the build string NEEDS to be prefixed with the platform value. If the user forgets about this, it will result in a filename collision. This should be checked by the linter, at the very least. I am unaware if sharing a hash is problematic for other cases.

File: src/maintainer/knowledge_base.rst
Comment (jakirkham) [inline src/maintainer/knowledge_base.rst:1089]: Would it be possible to avoid manipulating the `string` and instead rely on different hashes (note Jaime is mentioning something similar elsewhere)?

The main issue with `string` is pretty finicky. Users need to be very careful to replicate conda-build's behavior and if they missing something they could wind up with duplicate packages (which sometimes can both be accepted by Anaconda.org with neither being downloadable).

File: src/maintainer/knowledge_base.rst
Comment (jakirkham) [reply src/maintainer/knowledge_base.rst]: I share this concern and would prefer not using `string`. Explained in more detail below ( https://github.com/conda-forge/conda-forge.github.io/pull/1839#discussion_r1008679066 )

Comment (isuruf) [reply src/maintainer/knowledge_base.rst]: See https://github.com/conda/conda-build/pull/4606

File: src/maintainer/knowledge_base.rst
Comment (jaimergp) [reply src/maintainer/knowledge_base.rst:1089]: This will be needed until https://github.com/conda/conda-build/pull/4606 lands and a new conda-build release is cut.

Is it ok to merge as is or would you rather have a little admonition mentioning this caveat?

Comment (jakirkham) [reply src/maintainer/knowledge_base.rst:1089]: My understanding from the thread above is using `conda_build_config.yaml` meant we would not need this.

Fixing conda-build to not need either is certainly better. That said, between setting a `string` and using `conda_build_config.yaml`, would prefer the latter (fewer foot guns).

Comment (jaimergp) [reply src/maintainer/knowledge_base.rst:1089]: I am partial to `conda_build_config.yaml` too as we transition to the new conda-build version.

**Comment by jaimergp** [comment]
After a conversation with @viniciusdc, we realized a simplified version of this approach can also be used to skip OS builds. A common use case that requires non-noarch is to skip Windows, so we can just add a `__unix` dependency to achieve the same result under a `noarch` scheme.

**Comment by BastianZim** [comment]
Is everyone happy with merging this? Would be good to have it in the docs.

File: src/maintainer/knowledge_base.rst
Comment (BastianZim) [inline src/maintainer/knowledge_base.rst:1128]: I'm getting a linter error with this: https://github.com/conda-forge/poetry-feedstock/pull/80#issuecomment-1354686510
Commit: https://github.com/conda-forge/poetry-feedstock/pull/80/commits/e2abf93666388971b43d25581a84b3356a2c8d41

Comment (jaimergp) [reply src/maintainer/knowledge_base.rst:1128]: Hm, we might need to run `conda smithy recipe-lint` locally to see the actual error.

**Comment by jaimergp** [comment]
> Is everyone happy with merging this? Would be good to have it in the docs.

We need to agree on the "better" solution to avoid the hash collision. See discussion [here](https://github.com/conda-forge/conda-forge.github.io/pull/1839#discussion_r1008679066).

File: src/maintainer/knowledge_base.rst
Comment (BastianZim) [reply src/maintainer/knowledge_base.rst:1128]:
```python
Traceback (most recent call last):
  File "/opt/homebrew/Caskroom/mambaforge/base/bin/conda-smithy", line 10, in <module>
    sys.exit(main())
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/cli.py", line 670, in main
    args.subcommand_func(args)
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/cli.py", line 514, in __call__
    lints, hints = lint_recipe.main(
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/lint_recipe.py", line 978, in main
    content = render_meta_yaml("".join(fh))
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/utils.py", line 90, in render_meta_yaml
    content = env.from_string(text).render(context)
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/jinja2/environment.py", line 1301, in render
    self.environment.handle_exception()
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/jinja2/environment.py", line 936, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "<template>", line 16, in top-level template code
TypeError: 'str' object is not callable
```
This is the full error.

Comment (jaimergp) [reply src/maintainer/knowledge_base.rst:1128]: Ah, I think we need to split `SUBDIR` in a different `set` expression. 
See https://stackoverflow.com/questions/30515456/split-string-into-list-in-jinja

So we should have:

```yaml
{% set os_prefix, arch_suffix = SUBDIR.split('-', 1) %}
string: "{{ os_prefix }}_pyh{{ PKG_HASH }}_{{ PKG_BUILDNUM }}"
```

Comment (BastianZim) [reply src/maintainer/knowledge_base.rst:1128]: Hmm, still getting the same error:

```python
Traceback (most recent call last):
  File "/opt/homebrew/Caskroom/mambaforge/base/bin/conda-smithy", line 10, in <module>
    sys.exit(main())
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/cli.py", line 670, in main
    args.subcommand_func(args)
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/cli.py", line 514, in __call__
    lints, hints = lint_recipe.main(
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/lint_recipe.py", line 978, in main
    content = render_meta_yaml("".join(fh))
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/utils.py", line 90, in render_meta_yaml
    content = env.from_string(text).render(context)
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/jinja2/environment.py", line 1301, in render
    self.environment.handle_exception()
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/jinja2/environment.py", line 936, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "<template>", line 3, in top-level template code
TypeError: 'str' object is not callable
```

Recipe:

```yaml
{% set name = "poetry" %}
{% set version = "1.3.1" %}
{% set os_prefix, arch_suffix = SUBDIR.split('-', 1) %}

package:
  name: {{ name|lower }}
  version: {{ version }}

source:
  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/poetry-{{ version }}.tar.gz
  sha256: fde98462ad5dc9879393157da93092206a3411117e25a4761a41c6d08f31aea8

build:
  noarch: python
  number: 1
  # You NEED to include the platform in the build string to avoid hash collisions
  string: "{{ os_prefix }}_pyh{{ PKG_HASH }}_{{ PKG_BUILDNUM }}"
  script: {{ PYTHON }} -m pip install . -vv
  entry_points:
    - poetry = poetry.console.application:main
```

File: src/maintainer/knowledge_base.rst
Comment (dopplershift) [inline src/maintainer/knowledge_base.rst:1054]:
```suggestion
2.  ``conda-forge.yml``'s :ref:`noarch_platforms` option.
```

Comment (jaimergp) [reply src/maintainer/knowledge_base.rst]: Good catch, thanks!

File: src/maintainer/knowledge_base.rst
Comment (jaimergp) [reply src/maintainer/knowledge_base.rst:1128]: Ugh, turns out `SUBDIR` is not defined. But `target_platform` works:

```yaml
  string: "{{ target_platform.split('-')[0] }}_pyh{{ PKG_HASH }}_{{ PKG_BUILDNUM }}"
```

Still, a bit ugly...

Comment (jaimergp) [reply src/maintainer/knowledge_base.rst]: Changed the PR so it's explicit about the platforms and doesn't use that much Jinja in the string.

File: src/maintainer/knowledge_base.rst
Comment (hmaarrfk) [inline src/maintainer/knowledge_base.rst:1095]:
```suggestion
    host:
      - python >=3.7
      # ...
    run:
      - python >=3.7
      - numpy
```

don't we want to encourage lower bounds on noarch packages?

**Review by hmaarrfk — COMMENTED** [review]
thanks using this now.

**Comment by jaimergp** [comment]
The conda-build PR we were waiting for was merged recently and should be available in the upcoming conda-build 3.25, which means we will be able to merge this longstanding PR once and for all!

**Comment by asmeurer** [comment]
The linter will need to be updated to recognize this approach. See for instance https://github.com/conda-forge/sympy-feedstock/pull/52.

**Comment by jaimergp** [comment]
@asmeurer, it was updated a while ago (see https://github.com/conda-forge/conda-smithy/pull/1682), but only for simple platform selectors (e.g. `# [win]`) not custom ones like in your example (`# [with_gmpy]`).

**Comment by mikemhenry** [comment]
Just a friendly bump, I was looking for this recently and it would be great for the documentation to be easier to find, any blockers left?

**Comment by jaimergp** [comment]
This is ready to go from my side. @conda-forge/core if someone has the time to click merge :) 🙏

**Review by xhochy — APPROVED** [review]

**Pull request merged**
